### PR TITLE
Allow User Registration and Authentication behavior configurable

### PR DESCRIPTION
**Pull request by OmolemoBlessingLethuloe**

### CODE REVIEW

**Comment by OmolemoBlessingLethuloe** [comment]
Related Issue: https://github.com/shesha-io/shesha-framework/issues/2073

File: shesha-reactjs/src/providers/auth/index.tsx [hunk 1]
```diff
@@ -347,12 +347,21 @@ const AuthProvider: FC<PropsWithChildren<IAuthProviderProps>> = ({
...
+    return response;
+  };
+
   const loginUserAsync = (loginFormData: ILoginForm) =>
     new Promise((resolve, reject) => {
       dispatch((dispatchThunk, getState) => {
         dispatchThunk(loginUserAction()); // We just want to let the user know we're logging in
 
         loginUserHttp(loginEndpoint, loginFormData)
+          .then((response) => checkRegistrationCompletion(response))
```
Comment (IvanIlyichev) [inline shesha-reactjs/src/providers/auth/index.tsx]: Hi @OmolemoBlessingLethuloe. This chain of calls may be broken by the redirect. Please review the code and test it  thoroughly, looks like some of handlers in the chain should be merged with explicit branching

File: shesha-core/src/Shesha.Framework/Domain/ShaUserRegistration.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+        /// <summary>
+        /// User name or email address
+        /// </summary>
+        public virtual string UserNameOrEmailAddress { get; set; }
+
+        /// <summary>
+        /// Module name
+        /// </summary>
+        public virtual string AdditionalRegistrationInfoFormModule { get; set; }
```
Comment (IvanIlyichev) [inline shesha-core/src/Shesha.Framework/Domain/ShaUserRegistration.cs]: Hi @OmolemoBlessingLethuloe. I see you use two columns for module and form name. Please change it to use **FormIdentifier** type instead like this: ```public virtual FormIdentifier AdditionalRegistrationForm { get; set; }```

File: shesha-core/src/Shesha.Application/Authorization/TokenAuthController.cs [hunk 3]
```diff
@@ -60,19 +62,33 @@ public TokenAuthController(
...
+        [HttpPost]
+        public async Task<IActionResult> Authenticate([FromBody] AuthenticateModel model)
+        {
+            // Check for user registration status
+            var registration = await _userRegistration.FirstOrDefaultAsync(e => e.UserNameOrEmailAddress == model.UserNameOrEmailAddress);
+
+            if (registration != null && !registration.IsComplete)
+            {
+                // Return a custom result indicating a client-side redirect
+                var redirectUrl = $"/no-auth/{registration.AdditionalRegistrationInfoFormModule}/{checkRegistration.AdditionalRegistrationInfoFormName}";
```
Comment (IvanIlyichev) [inline shesha-core/src/Shesha.Application/Authorization/TokenAuthController.cs]: Hi @OmolemoBlessingLethuloe. **no-auth** route is specific to the front-end application, it should be moved to the front-end specific setting with **no-auth** as a default

File: shesha-core/src/Shesha.Framework/Otp/IOtpManager.cs [hunk 4]
```diff
@@ -9,17 +9,17 @@ public interface IOtpManager
         /// <summary>
         /// Send one-time-pin
         /// </summary>
-        Task<ISendPinResponse> SendPinAsync(ISendPinInput input);
+        Task<ISendPinResponse> SendPinAsync(SendPinInput input);
```
Comment (IvanIlyichev) [inline shesha-core/src/Shesha.Framework/Otp/IOtpManager.cs:12]: Hi @OmolemoBlessingLethuloe. What was a reason of removal of interface usage here? If it's not required for your task please revert the change, I know one of devs are busy with enhancements of the OTP functionality and this change may complicate merge. Initially interfaces used here to make services more generic

**Review by IvanIlyichev — CHANGES_REQUESTED** [review]
Hi @OmolemoBlessingLethuloe. Back-end build fails and I can't finish review in details, but I added some inline comments. Please fix build and review them. Additional notes:
1. Replace usage of `window.location.search` with `query`
2. Replace usage of `globalState` with application context or page context, contact Alex for details
3. Avoid usage of field names when access setting fields, use dot notation instead. Even if we have no intellisense support there currently we'll improve it in future
4. Remove all console.log from built in forms included into your package

File: shesha-core/src/Shesha.Application/UserManagements/CreatePersonAccountDto.cs [hunk 5]
```diff
@@ -9,8 +9,8 @@ namespace Shesha.Persons
     [AutoMapTo(typeof(User), typeof(Person))]
     public class CreatePersonAccountDto
     {
-        [Required(AllowEmptyStrings = false)]
-        [MinLength(5)]
+        //[Required(AllowEmptyStrings = false)]
+        //[MinLength(5)]
```
Comment (AlexStepantsov) [inline shesha-core/src/Shesha.Application/UserManagements/CreatePersonAccountDto.cs]: Why did you comment this?

File: shesha-core/src/Shesha.NHibernate/Migrations/M20240902170200.cs
Comment (AlexStepantsov) [inline shesha-core/src/Shesha.NHibernate/Migrations/M20240902170200.cs:1]: Why didn't you add the insertion of registration records for existing Persons? Now, a User Registration will always be created for new Person. So you need to add migration to add User Registration and for existing Persons

**Review by AlexStepantsov — CHANGES_REQUESTED** [review]

**Review by AlexStepantsov — APPROVED** [review]

**Review by AlexStepantsov — APPROVED** [review]

**Review by AlexStepantsov — APPROVED** [review]

**Pull request merged**